File: takkernel/engine/src/main/cpp/elevation/AbstractElevationData.cpp
#include "elevation/AbstractElevationData.h"

using namespace TAK::Engine;
using namespace TAK::Engine::Core;
using namespace TAK::Engine::Elevation;

AbstractElevationData::AbstractElevationData(int model, Port::String type, double resolution) NOTHROWS :
    model(model),
    type(type),
    resolution(resolution)
{}

AbstractElevationData::~AbstractElevationData() NOTHROWS
{
    // Nothing to do. Defined because the base method is defined as virtual.
}

int AbstractElevationData::getElevationModel() NOTHROWS 
{
    return model;
}

Util::TAKErr AbstractElevationData::getType(Port::String &datatype) NOTHROWS
{
    datatype = Port::String(type);

    return Util::TE_Ok;
}

double AbstractElevationData::getResolution() NOTHROWS
{
    return resolution;
}

Util::TAKErr AbstractElevationData::getElevation(
    double *elevations,
    Port::Collection<GeoPoint2>::IteratorPtr &points,
    const Hints &hint) NOTHROWS
{
    // Note: Hint is ignored in this implementation, which is based on the ATAK implementation.

    Util::TAKErr code(Util::TE_Ok);
    std::size_t idx = 0;
    do {
        GeoPoint2 point;
        code = points->get(point);
        TE_CHECKBREAK_CODE(code);

        double el(NAN);
        getElevation(&el, point.latitude, point.longitude);
        elevations[idx++] = el;

        code = points->next();
        TE_CHECKBREAK_CODE(code);
    } while (true);

    if (code == Util::TE_Done)
        code = Util::TE_Ok;
    TE_CHECKRETURN_CODE(code);

    return code;
}

File: takkernel/engine/src/main/java/com/atakmap/map/elevation/AbstractElevationData.java
package com.atakmap.map.elevation;

import java.util.Iterator;

import com.atakmap.coremap.maps.coords.GeoPoint;

public abstract class AbstractElevationData implements ElevationData {

    protected final int model;
    protected final String type;
    protected final double resolution;
    
    protected AbstractElevationData(int model, String type, double resolution) {
        this.model = model;
        this.type = type;
        this.resolution = resolution;
    }
    
    @Override
    public final int getElevationModel() {
        return this.model;
    }

    @Override
    public final String getType() {
        return this.type;
    }

    @Override
    public double getResolution() {
        return this.resolution;
    }

    @Override
    public abstract double getElevation(double latitude, double longitude);

    @Override
    public void getElevation(Iterator<GeoPoint> points, double[] elevations, Hints ignored) {
        int index = 0;
        GeoPoint point;
        while(points.hasNext()) {
            point = points.next();
            elevations[index++] = this.getElevation(point.getLatitude(), point.getLongitude());
        }
    }
}
